ts/packages/ts/src/ConfigSchema.ts

```diff
@@ -1,463 +1,496 @@
-              suggestion: `value ${cfgu.default} must be one of ${_.map(options, (option) => `'${option}'`).join(',')}`,
-            }),
-          );
-        }
-        if (cfgu.default && (cfgu.required || cfgu.template)) {
-          throw new Error(
-            ERR(`invalid default property`, {
-              location: [...scopeLocation, key, 'default'],
-              suggestion: `default mustn't set together with required or template properties`,
-            }),
-          );
+          throw error;
         }
+      });
+  }
 
-        if (cfgu.default && !ConfigSchema.CFGU.TESTS.VAL_TYPE[type]?.({ ...cfgu, value: cfgu.default })) {
-          throw new Error(
-            ERR(`invalid default property`, {
-              location: [...scopeLocation, key, 'default'],
-              suggestion: `value "${cfgu.default}" must be of type "${type}"`,
-            }),
-          );
-        }
+  static async fromFile(path: string, contents: string): Promise<ConfigSchema> {
+    const error = new ConfigError(
+      'invalid config schema path',
+      `path extension must be ${ConfigSchema.EXT.join('|')}`,
+      [['ConfigSchema', path]],
+    );
 
-        const isInvalidDepends =
-          cfgu.depends && (_.isEmpty(cfgu.depends) || cfgu.depends.some((depend) => !NAME(depend)));
-        if (isInvalidDepends) {
-          throw new Error(
-            ERR(`invalid depends property`, {
-              location: [...scopeLocation, key, 'depends'],
-              suggestion: `depends is empty or contain reserved words`,
-            }),
-          );
-        }
+    const splittedPath = path.split('.');
 
-        // todo: this is a "weak" validation and NAME() util collides with CONFIGU_SET.[prop]
-        // const isInvalidTemplate =
-        //   cfgu.template && TMPL.parse(cfgu.template).some((exp) => exp.type === 'name' && !NAME(exp.key));
-        // if (isInvalidTemplate) {
-        //   throw new Error(
-        //     ERR(`invalid template property`, {
-        //       location: [...scopeLocation, key, 'template'],
-        //       suggestion: `template is invalid or contain reserved words`,
-        //     }),
-        //   );
-        // }
-      });
+    const fileExt = splittedPath.pop();
+    if (!fileExt || !ConfigSchema.TYPES.includes(fileExt)) {
+      throw error;
+    }
 
-    return schemaContents;
-  }
-}
+    const cfguExt = splittedPath.pop();
+    if (cfguExt !== ConfigSchema.CFGU.NAME) {
+      throw error;
+    }
 
-export class InMemoryConfigSchema extends ConfigSchema {
-  constructor(
-    public contents: { [key: string]: Cfgu },
-    public name: string = '',
-  ) {
-    super(`${name}.cfgu.json`);
-  }
+    const schemaName = splittedPath.pop();
+    if (!schemaName) {
+      error.setHint(`name "${schemaName}" mustn't be empty`);
+      throw error;
+    }
+    if (!NAME(schemaName)) {
+      error.setHint(`name "${schemaName}" mustn't contain reserved words`);
+      throw error;
+    }
```

Comment:
This should be extracted and used by the constructor too, as it stands, schema names are only validated when loaded from a file and not by the constructor. It should be consistent.

Reply:
@RichardAkman and there the bug here with the path having `/` chars